CRM/Activityreport/Data.php

```diff
@@ -50,12 +161,63 @@ public static function get($offset = 0, $limit = 0, $multiValuesOffset = 0, $sta
       $params['activity_date_time'] = $activityDateFilter;
     }
 
-    $activities = civicrm_api3('Activity', 'get', $params);
+    while ($offset < $total) {
+      if ($offset) {
+        $offset--;
+      }
 
-    $formattedActivities = self::formatResult($activities['values']);
-    $result = self::splitMultiValues($formattedActivities, $offset, $multiValuesOffset);
+      $params['options']['offset'] = $offset;
 
-    return $result;
+      $activities = civicrm_api3('Activity', 'get', $params);
+
+      $formattedActivities = self::formatResult($activities['values']);
+
+      unset($activities);
+
+      while (!empty($formattedActivities)) {
+        $result = self::splitMultiValues($formattedActivities, $offset, $multiValuesOffset);
+
+        if ($result['info']['date'] !== $date) {
+          $page = 0;
+          $date = $result['info']['date'];
+        }
+
+        $cachedCount += self::cachePacket($result['data'], $date, $page++);
+
+        unset($result['data']);
+
+        $formattedActivities = array_slice($formattedActivities, $result['info']['nextOffset'] - $offset, NULL, TRUE);
+
+        $offset = $result['info']['nextOffset'];
+        $multiValuesOffset =  $result['info']['multiValuesOffset'];
+      }
+    }
+
+    return $cachedCount;
```

Comment:
So this method is actually doing 3 things:
- Fetch the data
- Split it into pages
- Store it in the cache

Ideally, each of these responsibilities should be in a separate method